Add climatology generation functionality

nikola-rados opened the pull request:
This PR introduces the `generate_climos` functionality from `climate-explorer-data-prep` as a `wps` process.  Given some `netCDF` file it will generate climatologies based on the options selected.  The process is meant to be used in `jupyter` with the `birdy` or `owslib` packages.  To test the functionality run the demo in `notebooks/`.

jameshiebert commented on thunderbird/processes/wps_generate_climos.py:
```diff
@@ -0,0 +1,296 @@
+        # loop over given climo, replace items ('historical', 'futures') with
+        # items from list
+        return list(
+            set(
```
I think you can use the set comprehension operator here: `{ x for x in iterable }`.

jameshiebert commented on thunderbird/processes/wps_generate_climos.py:
```diff
@@ -0,0 +1,296 @@
+            for attr in "time_resolution is_multi_year_mean".split():
+                output.append(f"{attr}: {getattr(input_file, attr)}")
+
+        filename = os.path.join(self.workdir, "dry.txt")
```
Does every process get its own unique workdir? And do you *have* to use a named file and return it?

Is it a possibility to use `tempfile.TemporaryFile(suffix='.txt', dir=self.workdir)` or `tempfile.NamedTemporaryFile(...)`?

jameshiebert replied:
Broader question: how and when do birds clean up their output?

nikola-rados replied:
This is a good question.  From what I've seen so far it seems like each process call creates a directory in `/tmp/`.  It also doesn't clean up after the process is done (at least on my local machine tests), I don't know how it behaves when running it from a docker container.  I imagine the process results would go down with the container but I can't say for sure just yet.

I can try changing to a tempfile and see what happens.

jameshiebert commented on tests/test_wps_generate_climos.py:
```diff
@@ -0,0 +1,58 @@
+    ),
+    [
+        ("mean", "6190", "yearly", "True", "True", "True", "False"),
+        ("std", "7100", "seasonal", "False", "False", "False", "True"),
```
Should probably have a test case for `resolutions="all"` here. And bonus... `resolutions={'all', 'yearly'}` in which case, I don't think your code will do the right thing :wink:

jameshiebert commented on thunderbird/processes/wps_generate_climos.py:
```diff
@@ -0,0 +1,296 @@
+
+    def format_resolutions(self, resolutions):
+        if "all" in resolutions:
+            return [
```
This is equivalent to `resolutions.remove('all')`, but I also don't think that will work if you have a case where someone supplies `'all'` as part of the set *without* the rest of the set. I.e. what happens if they submit `resolutions={'all', 'yearly'}`?

jameshiebert commented on thunderbird/processes/wps_generate_climos.py:
```diff
@@ -0,0 +1,296 @@
+        elif "netcdf" in request.inputs:
+            return request.inputs["netcdf"][0].file
+        else:
+            raise ProcessError("You must provide a data source")
```
What *was* provided? Can you add the provided inputs to the Exception data to help the user figure out what went wrong?

jameshiebert commented on thunderbird/processes/wps_generate_climos.py:
```diff
@@ -0,0 +1,296 @@
+        ) = self.collect_args(request)
+        filepath = self.get_filepath(request)
+
+        progress += 5
```
Do you have any better ways to measure progress here? This seems... dubious. Also, I'm pretty sure that not each independent path through this function will add up to 100.

nikola-rados replied:
I agree with you, it is not the best way to go about it.  I tried to write a class to make it cleaner but really it didn't help (if anything it made it worse).  Do you have any suggestions on how to improve this?

> Also, I'm pretty sure that not each independent path through this function will add up to 100

As long as the last line has the value `100` the process will be completed (although if you add up progress to `>= 100` it will send the "completed" message early, so still not great).
```python
response.update_status("Process Complete", 100)
```

jameshiebert commented:
PS The `cdo` process continues to annoyingly refuse to be killed by KeyboardInterrupt. Have you gotten anywhere with that?

nikola-rados commented on README.md:
```diff
@@ -0,0 +1,53 @@
+
+## Documentation
+
+Learn more about thunderbird in its official [documentation](https://thunderbird.readthedocs.io).
```
Those links were automatically created.  I was hoping to get to documentation in a later issue.

nikola-rados commented on README.md:
````diff
@@ -0,0 +1,53 @@
+
+Use the built-in `cli`:
+```
+(venv)$ thunderbird start
````
That is actually coming from the `ce-dataprep` package.  I've seen it come up every time I used the scripts from `climate-explorer-data-prep`.  I spoke to @corviday briefly about it and it didn't seem like an issue.  I suppose I could add `xarray` and see if the message goes away.

nikola-rados commented on notebooks/README.md:
````diff
@@ -0,0 +1,22 @@
+(demo_venv)$ jupyter labextension install @jupyter-widgets/jupyterlab-manager
+```
+
+**Note:** You may need to update `nodejs` in order to see the progress bars.
````
What exactly needed to be configured with `nvm`?

jameshiebert replied:
```bash
$ nvm install 10
$ nvm use 10
```

But it probably depends on what the exact requirements are.

nikola-rados commented on tests/common.py:
```diff
@@ -2,9 +2,21 @@
 
+import os
+
 VERSION = "1.0.0"
```
I'm not too sure to be honest.  I think it is supposed to mimic the version of the application or something.  It is set but not really used anywhere, we could probably get rid of it.

nikola-rados commented on tests/common.py:
```diff
@@ -2,9 +2,21 @@
+
+TESTDATA = {
+    "test_local_nc": "file:///{}".format(resource_file("test.nc")),
+    "test_opendap": "http://test.opendap.org:80/opendap/netcdf/examples/sresa1b_ncar_ccsm3_0_run1_200001.nc",  # noqa
```
Ah, yes I forgot to remove that.

nikola-rados commented on thunderbird/processes/wps_generate_climos.py:
```diff
@@ -0,0 +1,296 @@
+                "Output",
+                abstract="Collection of climatoligical files",
+                as_reference=True,
+                supported_formats=[FORMATS.META4],
```
You can see it [here](https://pywps.readthedocs.io/en/master/_modules/pywps/inout/formats.html).  It was the suggested format for processes with multiple outputs as shown in [emu](https://github.com/bird-house/emu/blob/master/emu/processes/wps_multiple_outputs.py).  I have other `birds` create `zip` or `tar` files instead.  That is something we could consider.

nikola-rados commented:
The folks who make `cdo` have no done anything about my issue.  I'll find some time to put up a PR and hope that they accept it.

> The setup.py file doesn't do the right thing with the requirements... it is not supposed pin the requirements, but it's using requirements.txt verbosely, which isn't right. I'd suggest changing it to enumerate the non-pinned deps.

I'm not sure I understand what exactly you mean by this. Would you be able to explain how/where the `setup.py` file is doing this?

nikola-rados commented on tests/test_describe.py:
```diff
@@ -0,0 +1,16 @@
+from owslib.wps import WebProcessingService
+
+
+def test_describe():
```
This test is just creating a test instance of thunderbird... Really I think the job of this test has been replaced by the notebook demo.  I think I'll remove it.

nikola-rados commented on thunderbird/processes/wps_generate_climos.py:
```diff
@@ -0,0 +1,296 @@
+        super(GenerateClimos, self).__init__(
+            self._handler,
+            identifier="generate_climos",
+            version="0.8.0",
```
As it turns out python doesn't like the `-` in that name.  `dp` doesn't have a `__version__` attribute so I may just leave it hard coded for now.

nikola-rados commented:
@jameshiebert I've addressed the feedback that I was able to. I still have the same questions lingering from the comment above. Let me know if there is anything else you would like changed.
